### ISSUE TITLE
Enable passing credentials to fetch?

### ISSUE THREAD

**Issue by itrew** [Esri/arcgis-rest-js#443]
Continuing the discussion started in #290 after it was closed with the implementation of custom `headers`.

Right now, the `request` function has a hard-coded `credentials` option that only passes credentials when the resource has the same origin as the request.

https://github.com/Esri/arcgis-rest-js/blob/462114610b9ca5750a933f297764ed46bc5c56eb/packages/arcgis-rest-request/src/request.ts#L147-L151

While that works great a lot of the time, there will be instances where the REST endpoint ends up being on another origin. Say if someone has an application hosted on one machine (`example.com`) but wants to use IWA to access an ArcGIS Server resource that is on another machine (`geoexample.com`), it would require the `credentials: 'include'` option to be set with fetch. Right now, the only way around that is to pass a custom fetch implementation to `request` like below.

```javascript
// Current workaround
import { request } from '@esri/arcgis-rest-request';

function fetchWithCredentials(url, opts = {}) {
  return fetch(
    url,
    Object.assign(opts, {
      credentials: 'include',
    }),
  );
}

function customFetch(url, opts) {
  return Object.assign(fetchWithCredentials.bind(null, url, opts), fetch)();
}

request('some url', {
  fetch: customFetch
})
```

It would be nice to be able to add the ability to overwrite the default `credentials` property. It would require adding a property to `IRequestOptions` and slightly modifying the `request` function.

```typescript
// packages/arcgis-rest-request/src/request.ts
// Changes to IRequestOptions
export interface IRequestOptions {
  // ...Already existing properties
  /**
   * Should the user agent send cookies from the other domain for credentials.
   */
  credentials?: "omit" | "same-origin" | "include"
}

// Changes to request()
export function request(
  url: string,
  requestOptions: IRequestOptions = { params: { f: "json" } }
): Promise<any> {
  const options: IRequestOptions = {
    httpMethod: "POST",
    // Set it to the default. This may not even be necessary as it is the default for fetch anyways.
    credentials: "same-origin"
    ...requestOptions
  };
  const { httpMethod, authentication, fetchCredentials } = options;
  const fetchOptions: RequestInit = {
    method: httpMethod,
    // ensures behavior mimics XMLHttpRequest. needed to support sending IWA cookies
    credentials: fetchCredentials
  };
}
```

**Comment by jgravois** [comment]
thanks @itrew. i'd like to tackle this _after_ #209 is complete.

**Comment by jgravois** [comment]
breadcrumb: https://github.com/Esri/arcgis-rest-js/issues/290#issuecomment-455849033

**Comment by tomwayson** [comment]
@itrew I believe this was resolved in #831, can you please confirm?
